include/platform/mir/renderer/sw/pixel_source.h

```diff
@@ -30,6 +35,74 @@ namespace renderer
 namespace software
 {
 
+class BufferDescriptor
+{
+public:
+    virtual ~BufferDescriptor() = default;
+
+    virtual MirPixelFormat format() const = 0;
+    virtual geometry::Stride stride() const = 0;
+    virtual geometry::Size size() const = 0;
+};
+
+template<typename T>
+class Mapping : public BufferDescriptor
+{
+public:
+    /**
+     * Destroy the mapping of the buffer.
+     *
+     * This does not destroy the underlying buffer, only the CPU-accessible
+     * view of it.
+     */
+    virtual ~Mapping() = default;
+
+    virtual T* data() = 0;
+    virtual size_t len() const = 0;
+};
+
+/**
+ * A Buffer that can be mapped into CPU-accessible memory and directly written to.
+ */
+class WriteMappableBuffer
+{
+public:
+    virtual ~WriteMappableBuffer() = default;
+
+    virtual std::unique_ptr<Mapping<unsigned char>> map_writeable() = 0;
+};
+
+/**
+ * A Buffer that can be mapped into CPU-accessible memory and directly read from.
+ */
+class ReadMappableBuffer
+{
+public:
+    virtual ~ReadMappableBuffer() = default;
+    virtual std::unique_ptr<Mapping<unsigned char const>> map_readable() = 0;
+};
+
+class ReadTransferableBuffer : public virtual BufferDescriptor
+{
+public:
+    virtual ~ReadTransferableBuffer() = default;
+
+    virtual void transfer_from_buffer(unsigned char* destination) const = 0;
+};
+
+class WriteTransferableBuffer : public virtual BufferDescriptor
+{
+public:
+    virtual ~WriteTransferableBuffer() = default;
+
+    virtual void transfer_into_buffer(unsigned char const* source) = 0;
+};
+
+auto as_read_mappable_buffer(
+    std::shared_ptr<graphics::Buffer> buffer) -> std::shared_ptr<ReadMappableBuffer>;
+auto as_write_mappable_buffer(
+    std::shared_ptr<graphics::Buffer> buffer) -> std::shared_ptr<WriteMappableBuffer>;
+
```

Comment:
Do these need to take `buffer` by value?

Reply:
They don't *have* to, but they always consume `buffer` so there's no advantage to passing in a `shared_ptr<> const&`.

clang-tidy started to suggest passing-by-value in such cases, and I don't think it's wrong.